crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs

```diff
@@ -2785,47 +2788,74 @@ impl Analyzer<'_, '_> {
-                            }
-                            .into());
-                        }
+    pub(super) fn assign_to_intrinsic(&mut self, data: &mut AssignData, to: &StringMapping, r: &Type, opts: AssignOpts) -> VResult<()> {
+        match r.normalize() {
+            Type::Keyword(KeywordType {
+                kind: TsKeywordTypeKind::TsAnyKeyword,
+                ..
+            })
+            | Type::Keyword(KeywordType {
+                kind: TsKeywordTypeKind::TsNeverKeyword,
+                ..
+            }) => {
+                return Ok(());
+            }
+            Type::Keyword(KeywordType {
+                kind: TsKeywordTypeKind::TsStringKeyword,
+                ..
+            }) => {
+                if self.ctx.in_actual_type {
+                    return Ok(());
+                }
+
+                // declare var x: Uppercase<string>
+                if let PatMode::Decl = self.ctx.pat_mode {
+                    if to.type_args.params[0].is_str() {
+                        return Ok(());
                     }
                 }
-                IntrinsicKind::Lowercase => {
-                    if let Some(value) = &value.raw {
-                        if value.to_lowercase() != **value {
-                            return Err(ErrorKind::AssignFailed {
-                                span: r.span(),
-                                left: box Type::StringMapping(to.clone()),
-                                right_ident: None,
-                                right: box r.clone(),
-                                cause: vec![],
-                            }
-                            .into());
-                        }
+
+                if self.ctx.in_return_arg {
+                    if to.type_args.params[0].is_str_like() {
+                        return Ok(());
                     }
                 }
-                IntrinsicKind::Capitalize => {
-                    if let Some(value) = &value.raw {
-                        let ch = value.chars().next();
 
-                        if let Some(ch) = ch {
-                            if !ch.is_uppercase() {
+                // TODO: Handle this better
+                let mut span = opts.span;
+                if span.is_dummy() {
+                    span = to.span();
+                }
+
+                return Err(ErrorKind::AssignFailed {
+                    span,
+                    left: box Type::StringMapping(to.clone()),
+                    right_ident: None,
+                    right: box r.clone(),
+                    cause: vec![],
+                }
+                .into());
+            }
+
+            Type::Lit(LitType {
+                lit: RTsLit::Str(str_lit), ..
+            }) => {
+                let type_param = &to.type_args.params[0];
+
+                if self.ctx.in_actual_type {
+                    return Ok(());
+                }
+
+                if type_param.is_alias() || type_param.is_ref_type() {
```

Comment:
This seems wrong. Shouldn't we normalize it instead?

Reply:
We could do something like
```rs
                if type_param.is_alias() || type_param.is_ref_type() {
                    return self.assign(to.span(), &mut Default::default(), type_param, r);
                }
```

or were you thinking of calling `self.normalize` ?

Reply:
removed the if clause completely